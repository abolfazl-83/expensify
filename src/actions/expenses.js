import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => {
    return {
        type : 'ADD_EXPENSE',
        expense
    }
};
export const startAddExpense = ({description , amount = 0 , note = '' , createdAt = 0} = {}) => {
    return (dispatch , getState) => {
        const uid = getState().auth.uid;
        const expense = {
            description,
            amount,
            note,
            createdAt
        };
        return database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
            dispatch(addExpense({
                id : snapshot.key,
                ...expense
            }))
        })
    };
};
export const editExpense = (id , updates) => {
    return {
        type : 'EDIT_EXPENSE',
        id,
        updates
    }
};

export const startEditExpense = (id , updates) => {
    return (dispatch , getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id , updates))
        });
    }
};

export const removeExpense = (id) => {
    return {
        type : 'REMOVE_EXPENSE',
        id
    }
};

export const startRemoveExpense = (id) => {
    return (dispatch , getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense(id))
        })
    }
};

export const setExpenses = (expenses) => {
    return {
        type : 'SET_EXPENSES',
        expenses
    }
}

export const  startSetExpenses = () => {
    return (dispatch , getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((dataSnapshot) => {
            const expenses = [];

            dataSnapshot.forEach((childSnapshot) => {
                expenses.push({
                    id : childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        });
    }
};