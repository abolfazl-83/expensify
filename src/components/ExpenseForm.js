import React,{useState} from 'react';
import {SingleDatePicker} from 'react-dates';
import moment from 'moment';

export const ExpenseForm = ({onSubmit , expense}) => {
    const[description, setDescription] = useState(() => (expense) ? expense.description : '');
    const[amount, setAmount] = useState(() => (expense) ? (expense.amount / 100).toString() : '');
    const[note , setNote] = useState(() => (expense) ? expense.note : '');
    const[date, setDate] = useState(() => (expense) ? moment(expense.createdAt) : moment());
    const[focused, setFocused] = useState(false);
    const[error, setError] = useState('');
    
    const handleOnDescriptionChange = (e) => {
        const value = e.target.value;
        setDescription(value);
    };
    const handleOnAmountChange = (e) => {
        const value = e.target.value;
        if( !value || value.match(/^\d{1,}(\.\d{0,2})?$/)){
            setAmount(value);
        }
    };
    const handleOnNoteChange = (e) => {
        const value = e.target.value;
        setNote(value);
    }
    const handleOnDateChange = (date) => {
        setDate(date);
    };
    const handleOnFocusChanged = ({focused}) => {
        setFocused(focused)
    };
    const handleFormSubmit = (e) => {
        e.preventDefault();
        if(!(description && amount)){
            setError('description and amount are required.');
        } else {
            onSubmit({
                description : description,
                amount : parseFloat(amount * 100),
                note : note,
                createdAt : date.valueOf()
            })
        }
    };
    return (
        <form
        className = 'form'
        onSubmit = {handleFormSubmit}
        >
            {error && <p className = 'form__error'>{error}</p>}
            <input 
            className = 'text-input'
            type = 'text'
            placeholder = 'description...'
            name = 'descriptionInput'
            value = {description}
            onChange = {handleOnDescriptionChange}
            /> 
            <input
            className = 'text-input' 
            type = 'text'
            placeholder = 'amount...'
            name = 'amountInput'
            value = {amount}
            onChange = {handleOnAmountChange}
            /> 
            <SingleDatePicker
            id = 'singleDatePicker' 
            date = {date}
            onDateChange = {handleOnDateChange}
            focused = {focused}
            onFocusChange = {handleOnFocusChanged}
            showClearDate = {true}
            isOutsideRange = {() => false}
            numberOfMonths = {1}
            hideKeyboardShortcutsPanel = {true}
            readOnly = {true}
            />
            <textarea
            className = 'textarea'
            placeholder = 'note...'
            name = 'noteTextarea'
            value = {note}
            onChange = {handleOnNoteChange}
            >
            </textarea>
            <div>
                <button className = 'button'>{(expense) ? 'Edit Expense' : 'Add Expense'}</button>
            </div>                      
        </form>
    );
};

export default ExpenseForm;