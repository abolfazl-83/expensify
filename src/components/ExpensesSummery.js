import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import numeral from 'numeral';
import expensesSummery from '../selectors/expensesSummery';
import visibleExpenses from '../selectors/visibleExpenses';


export const ExpensesSummery = ({expensesSummeryResult : {expensesCount , expensesTotal}}) => {

    const expensesWord = (expensesCount === 1) ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div className = 'page-header'>
            <div className = 'content-container'>
                <h1 className = 'page-header__title'>viewing <span>{expensesCount}</span> {expensesWord} totalling <span>{formattedExpensesTotal}</span></h1>
                <div className = 'page-header__actions'>
                    <Link
                    className = 'button'
                    to = '/create'
                    >Add Expense</Link>
                </div>
            </div>
        </div>
    );
}

const mapStateToProps = (state) => {
    return {
        expensesSummeryResult : expensesSummery(visibleExpenses(state.expenses, state.filters))
    }
};

export default connect(mapStateToProps)(ExpensesSummery);