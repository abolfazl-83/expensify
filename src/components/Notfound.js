import React from 'react';

export const Notfound = ({history}) => {
    return (
        <React.Fragment>
            <h2>the page you are looking for, does not exist in our servers. if u disagree, contact our support staff.</h2>
            <button onClick = {() => {
                history.push('/');
            }}>Go Home</button>
        </React.Fragment>
    );
};

export default Notfound;