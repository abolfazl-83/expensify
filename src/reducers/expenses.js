const expensesReducerDefaultValue = [];
const expensesReducer = (state = expensesReducerDefaultValue,action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => {
                return expense.id !== action.id;
            });
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id !== action.id){
                    return expense;
                } else {
                    return {
                        ...expense,
                        ...action.updates
                    };
                };
            });
        case 'SET_EXPENSES':
            return [
                ...action.expenses
            ]
        default : return state;
    }
};

export default expensesReducer;