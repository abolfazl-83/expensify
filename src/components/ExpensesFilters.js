import React, {useState} from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {sortByDate , sortByAmount , setFilterText , setStartDate , setEndDate} from '../actions/filters';


export const ExpensesFilters = ({filters : {text , sortBy , startDate , endDate} ,setFilterText, sortByAmount, sortByDate , setStartDate , setEndDate}) => {
    const[focusedInput , setFocusedInput] = useState(null);

    const handleOnTextChange = (e) => {
        const value = e.target.value;
        setFilterText(value);
    };
    const handleOnSelectChange = (e) => {
        if(e.target.value === 'amount') {
            sortByAmount();
        } else {
            sortByDate()
        }
    }
    const handleOnDatesChange = ({startDate , endDate}) => {
        setStartDate(startDate);
        setEndDate(endDate)
    }
    const handleOnFocusChange = (focusedInput) => {
        setFocusedInput(focusedInput)
    }
    return (
        <div className = 'content-container'>
            <div className = 'input-group'>
                <div className = 'input-group__item'>
                    <input 
                        type = 'text'
                        className = 'text-input'
                        placeholder = 'search your expense...'
                        name = 'filterTextInput'
                        value = {text}
                        onChange = {handleOnTextChange}
                    />
                </div>
                <div className = 'input-group__item'>
                    <select
                    className = 'select'
                    value = {sortBy}
                    onChange = {handleOnSelectChange}
                    >
                         <option>date</option>
                         <option>amount</option>
                    </select>
                </div>
                <div className = 'input-group__item'>
                    <DateRangePicker 
                     startDate = {startDate}
                     endDate = {endDate}
                     onDatesChange = {handleOnDatesChange}
                     onFocusChange = {handleOnFocusChange}
                     focusedInput = {focusedInput}
                     numberOfMonths = {1}
                     hideKeyboardShortcutsPanel = {true}
                     isOutsideRange = {() => false}
                     readOnly = {true}
                     showClearDates = {true}
                     startDateId = '12345'
                     endDateId = '98765'
                    />                    
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        filters : state.filters
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        setFilterText : (text) => dispatch(setFilterText(text)),
        sortByDate : () => dispatch(sortByDate()),
        sortByAmount : () => dispatch(sortByAmount()),
        setStartDate : (startDate) => dispatch(setStartDate(startDate)),
        setEndDate : (endDate) => dispatch(setEndDate(endDate))
    }
};

export default connect(mapStateToProps , mapDispatchToProps)(ExpensesFilters);