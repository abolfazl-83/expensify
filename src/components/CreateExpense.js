import React from 'react';
import {connect} from 'react-redux';
import {startAddExpense} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export const CreateExpense = ({startAddExpense , history}) => {
    const handleOnSubmit = ({ description , amount , note , createdAt}) => {
        startAddExpense({
            description,
            note,
            amount,
            createdAt
        })
        history.push('/')
    };
    return (
        <div>
            <div className = 'page-header'>
                <div className = 'content-container'>
                    <h1 className = 'page-heaer__title'>Add Expense</h1>
                </div>
            </div>
            <div className = 'content-container'>
                <ExpenseForm 
                 onSubmit = {handleOnSubmit}
                />
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => {
    return {
        startAddExpense : (expense) => dispatch(startAddExpense(expense))
    }
};

export default connect(undefined , mapDispatchToProps)(CreateExpense);