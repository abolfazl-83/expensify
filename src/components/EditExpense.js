import React from 'react';
import {connect} from 'react-redux';
import {startEditExpense , startRemoveExpense} from '../actions/expenses';
import ExpenseForm from '../components/ExpenseForm';

export const EditExpense = ({expense , startEditExpense , history, startRemoveExpense}) => {
    
    const handleOnSubmit = (updates) => {
        startEditExpense(expense.id , updates)
        history.push('/')
    }
    const handleOnRemove = () => {
        startRemoveExpense(expense.id);
        history.push('/')
    };
    return (
        <div>
            <div className = 'page-header'>
                <div className = 'content-container'>
                    <h1 className = 'page-header__title'>Edit Expense</h1>
                </div>
            </div>
            <div className = 'content-container'>
                <ExpenseForm 
                    expense = {expense}
                    onSubmit = {handleOnSubmit}
                />
                <button
                className = 'button button-secondary'
                onClick = {handleOnRemove}
                >
                Remove Expense
                </button>
            </div>
        </div>
    )
}
const mapStateToProps = (state, props) => {
    return {
        expense : state.expenses.find((expense) => {
            return expense.id === props.match.params.expenseId
        })
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        startEditExpense : (id , updates) => dispatch(startEditExpense(id , updates)),
        startRemoveExpense : (id) => dispatch(startRemoveExpense(id))
    }
};

export default connect(mapStateToProps , mapDispatchToProps)(EditExpense);