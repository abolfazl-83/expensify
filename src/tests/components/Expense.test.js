import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';
import {Expense} from '../../components/Expense';

let wrapper;

test('should render Expense correctly' , () => {
    wrapper = shallow(<Expense {...expenses[1]}/>)
    expect(wrapper).toMatchSnapshot();
});