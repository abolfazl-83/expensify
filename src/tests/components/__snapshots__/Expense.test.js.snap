// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense correctly 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    water payment
  </Link>
  <p>
    $200.00
  </p>
  <p>
    Saturday, December 27th 1969
  </p>
  <p>
    April water payment
  </p>
  <hr />
</div>
`;
